### PR TITLE
Don't send reminder notifications before suppressed ones

### LINKED ISSUE #8201: Dependency is causing notifications instead of preventing them
## Describe the bug

In our scenario we have a central Icinga 2 master which is monitoring departments on different locations. These departments always have a firewall, a switch, a hypervisor and multiple virtual machines. A monitoring object is created for every item and a dependency between switch and firewall, hypervisor and switch, vms and hypervisor. These dependencies should reduce the number of notifications and we tried to multiple settings with different number of check attempts, check and retry interval, but we can not get it reliably to not send unwanted notifications.

firewall_anonym
switch_anonym
esx_anonym

## To Reproduce

Configuration looks like this:

```
object Host "firewall" {
	address = "192.168.32.254"
	check_command = "hostalive"
	check_interval = 2m
	retry_interval = 1m
	max_check_attempts = 3
}

object Host "switch" {
	address = "192.168.32.222"
	check_command = "hostalive"
	check_interval = 3m
	retry_interval = 1m
	max_check_attempts = 3
}

object Host "esx" {
	address = "192.168.32.252"
	check_command = "hostalive"
	check_interval = 4m
	retry_interval = 1m
	max_check_attempts = 4
}

apply Dependency "switch-to-firewall" to Host {
        parent_host_name = "firewall"

        states = [ Up ]
	ignore_soft_states = false

	assign where host.name == "switch"
}

apply Dependency "hosts-to-switch" to Host {
        parent_host_name = "switch"

        states = [ Up ]
	ignore_soft_states = false

        assign where host.vars.name == "esx"
}
```

## Expected behavior

The dependency should only allow notifications if the systems are still down and not if the systems before them in the chain come back to OK.

## Screenshots

see above

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Version used (`icinga2 --version`): 2.11.4
* Operating System and version: RHEL 7

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #8201

### CODE REVIEW

**Comment by Al2Klimov** [comment]
In this case we still send notifications, but the suppressed ones and only because of disabled active checks. If they were enabled and the switch would recover soon, the suppressed problem notification would disappear as it would no longer apply and no reminders would be sent.

## Before

```
[2020-09-17 12:22:37 +0200] information/Notification: Completed sending 'Recovery' notification 'firewall!icingaadmin' for checkable 'firewall' and user 'icingaadmin' using command 'mail-host-notification'.
[2020-09-17 12:22:37 +0200] notice/Process: PID 70103 ('/Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/scripts/mail-host-notification.sh' '-4' '' '-6' '' '-b' '' '-c' '' '-d' '2020-09-17 12:22:37 +0200' '-l' 'firewall' '-n' 'firewall' '-o' ' ' '-r' 'icinga@localhost' '-s' 'UP' '-t' 'RECOVERY') terminated with exit code 0
[2020-09-17 12:22:39 +0200] notice/CheckerComponent: Pending checkables: 0; Idle checkables: 2; Checks/s: 0
[2020-09-17 12:22:39 +0200] notice/Dependency: Dependency 'switch!firewall' passed: Parent host 'firewall' matches state filter.
[2020-09-17 12:22:39 +0200] notice/NotificationComponent: Attempting to send reminder notification 'switch!icingaadmin'.
[2020-09-17 12:22:39 +0200] notice/Notification: Attempting to send reminder notifications of type 'Problem' for notification object 'switch!icingaadmin'.
[2020-09-17 12:22:39 +0200] debug/Notification: Type 'Problem', TypeFilter: Acknowledgement, Custom, DowntimeEnd, DowntimeRemoved, DowntimeStart, FlappingEnd, FlappingStart, Problem and Recovery (FType=32, TypeFilter=-1)
[2020-09-17 12:22:39 +0200] debug/Notification: State 'Down', StateFilter: Critical, Down, OK, Unknown, Up and Warning (FState=32, StateFilter=-1)
[2020-09-17 12:22:39 +0200] debug/Notification: User 'icingaadmin' notification 'switch!icingaadmin', Type 'Problem', TypeFilter: Acknowledgement, Custom, DowntimeEnd, DowntimeRemoved, DowntimeStart, FlappingEnd, FlappingStart, Problem and Recovery (FType=32, TypeFilter=-1)
[2020-09-17 12:22:39 +0200] debug/Notification: User 'icingaadmin' notification 'switch!icingaadmin', State 'Down', StateFilter: Critical, Down, OK, Unknown, Up and Warning (FState=32, StateFilter=-1)
[2020-09-17 12:22:39 +0200] information/Notification: Sending reminder 'Problem' notification 'switch!icingaadmin' for user 'icingaadmin'
```

## After

```
[2020-09-17 12:31:25 +0200] information/Notification: Completed sending 'Recovery' notification 'firewall!icingaadmin' for checkable 'firewall' and user 'icingaadmin' using command 'mail-host-notification'.
[2020-09-17 12:31:25 +0200] notice/Process: PID 71470 ('/Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/scripts/mail-host-notification.sh' '-4' '' '-6' '' '-b' '' '-c' '' '-d' '2020-09-17 12:31:25 +0200' '-l' 'firewall' '-n' 'firewall' '-o' ' ' '-r' 'icinga@localhost' '-s' 'UP' '-t' 'RECOVERY') terminated with exit code 0
[2020-09-17 12:31:25 +0200] notice/CheckerComponent: Pending checkables: 0; Idle checkables: 2; Checks/s: 0
[2020-09-17 12:31:25 +0200] notice/Dependency: Dependency 'switch!firewall' passed: Parent host 'firewall' matches state filter.
[2020-09-17 12:31:27 +0200] notice/Dependency: Dependency 'switch!firewall' passed: Parent host 'firewall' matches state filter.
[2020-09-17 12:31:27 +0200] notice/ApiListener: Relaying 'event::SendNotifications' message
[2020-09-17 12:31:27 +0200] information/Checkable: Checkable 'switch' has 1 notification(s). Checking filters for type 'Problem', sends will be logged.
[2020-09-17 12:31:27 +0200] notice/ApiListener: Relaying 'event::SetForceNextNotification' message
[2020-09-17 12:31:27 +0200] notice/Notification: Attempting to send notifications of type 'Problem' for notification object 'switch!icingaadmin'.
[2020-09-17 12:31:27 +0200] debug/Notification: Type 'Problem', TypeFilter: Acknowledgement, Custom, DowntimeEnd, DowntimeRemoved, DowntimeStart, FlappingEnd, FlappingStart, Problem and Recovery (FType=32, TypeFilter=-1)
[2020-09-17 12:31:27 +0200] debug/Notification: State 'Down', StateFilter: Critical, Down, OK, Unknown, Up and Warning (FState=32, StateFilter=-1)
[2020-09-17 12:31:27 +0200] debug/Notification: User 'icingaadmin' notification 'switch!icingaadmin', Type 'Problem', TypeFilter: Acknowledgement, Custom, DowntimeEnd, DowntimeRemoved, DowntimeStart, FlappingEnd, FlappingStart, Problem and Recovery (FType=32, TypeFilter=-1)
[2020-09-17 12:31:27 +0200] debug/Notification: User 'icingaadmin' notification 'switch!icingaadmin', State 'Down', StateFilter: Critical, Down, OK, Unknown, Up and Warning (FState=32, StateFilter=-1)
[2020-09-17 12:31:27 +0200] information/Notification: Sending 'Problem' notification 'switch!icingaadmin' for user 'icingaadmin'
```

**Comment by lippserd** [comment]
@Al2Klimov @N-o-X Are there any expected side effects here? I would suggest to include this fix in 2.11.6 and 2.12.1.

**Comment by N-o-X** [comment]
@lippserd nope, this should not have any side effects. I think we can safely add this to both versions.

File: lib/notification/notificationcomponent.cpp [hunk 1]
```diff
@@ -165,6 +165,11 @@ void NotificationComponent::NotificationTimerHandler()
 			if ((service && service->GetState() == ServiceOK) || (!service && host->GetState() == HostUp))
 				continue;
 
+			/* Don't send reminder notifications before initial ones. */
+			if (checkable->GetSuppressedNotifications() & NotificationProblem) {
+				continue;
+			}
```
Comment (N-o-X) [inline lib/notification/notificationcomponent.cpp:171]: Please use the same code style as the rest of the function.

**Review by N-o-X — CHANGES_REQUESTED** [review]

**Review by lippserd — APPROVED** [review]

**Pull request merged**